File: internal/cli/ctx/add.go
package ctx

import (
	"fmt"
	"os"
	"path/filepath"

	clisession "github.com/ntwrknrd/nssh/internal/cli/session"
	"github.com/ntwrknrd/nssh/internal/config"
	"github.com/ntwrknrd/nssh/internal/ui"
	"github.com/ntwrknrd/nssh/internal/vault"
	"github.com/spf13/cobra"
)

// NewAddCmd creates the ctx add command.
func NewAddCmd() *cobra.Command {
	var (
		sshConfig string
		domain    string
		username  string
		dryRun    bool
	)

	cmd := &cobra.Command{
		Use:   "add [name]",
		Short: "Create new context",
		Long: `Create new SSH context.

Prompts for context name, SSH config file, domain, and credentials interactively.
Use flags for non-interactive/scripted usage.

The --domain option enables automatic context selection when connecting
to hosts matching the domain suffix (e.g., --domain example.com will
auto-select this context for server.example.com).`,
		Args: cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			var name string
			if len(args) > 0 {
				name = args[0]
			}
			return runAdd(name, addFlags{
				sshConfig:    sshConfig,
				sshConfigSet: cmd.Flags().Changed("ssh-config"),
				domain:       domain,
				domainSet:    cmd.Flags().Changed("domain"),
				username:     username,
				usernameSet:  cmd.Flags().Changed("username"),
				dryRun:       dryRun,
			})
		},
	}

	cmd.Flags().StringVarP(&sshConfig, "ssh-config", "s", "", "SSH config include file")
	cmd.Flags().StringVarP(&domain, "domain", "d", "", "domain suffix for auto-selection")
	cmd.Flags().StringVarP(&username, "username", "u", "", "fallback username (password prompted securely)")
	cmd.Flags().BoolVar(&dryRun, "dry-run", false, "preview only")

	return cmd
}

type addFlags struct {
	sshConfig    string
	sshConfigSet bool
	domain       string
	domainSet    bool
	username     string
	usernameSet  bool
	dryRun       bool
}

func runAdd(name string, flags addFlags) error {
	mgr, err := clisession.NewManager(vault.Auto())
	if err != nil {
		return err
	}

	// Unlock vault if needed and TTY is available
	_ = clisession.TryUnlockIfTTY(mgr)

	ui.CommandStart("ADD CONTEXT")

	// Context name
	finalName := name
	if finalName == "" {
		if flags.dryRun {
			finalName = "<dry-run-context>"
		} else {
			finalName, err = ui.Input("Context name", "")
			if err != nil {
				ui.Abort("Canceled")
				ui.CommandEnd(ui.StatusAbort)
				return nil
			}
			if finalName == "" {
				ui.Error("Context name is required")
				ui.CommandEnd(ui.StatusError)
				return nil
			}
		}
	}

	// Check if context already exists (skip for dry-run)
	if !flags.dryRun {
		existing, err := mgr.GetContext(finalName)
		if err != nil {
			ui.Error("Failed to check context: %s", err)
			ui.CommandEnd(ui.StatusError)
			return nil
		}
		if existing != nil {
			ui.Error("Context '%s' already exists", finalName)
			ui.CommandEnd(ui.StatusError)
			return nil
		}
	}

	// SSH config file
	gitIncludeFile := flags.sshConfig
	if !flags.sshConfigSet {
		if flags.dryRun {
			gitIncludeFile = "<dry-run-file>"
		} else {
			// Get files already used by other contexts
			contexts, _ := mgr.ListContexts()
			usedFiles := make([]string, 0, len(contexts))
			for _, c := range contexts {
				usedFiles = append(usedFiles, c.GitIncludeFile)
			}

			defaultFilename := fmt.Sprintf("%s_hosts", finalName)
			gitIncludeFile, err = selectSSHConfigFile(defaultFilename, usedFiles...)
			if err != nil {
				ui.Abort("Canceled")
				ui.CommandEnd(ui.StatusAbort)
				return nil
			}
		}
	}

	// Domain
	domain := flags.domain
	if !flags.domainSet && !flags.dryRun {
		choice, err := ui.Select("Domain", []ui.SelectOption{
			{Label: "Skip (no domain)", Value: "skip"},
			{Label: "Set domain...", Value: "set"},
		})
		if err != nil {
			ui.Abort("Canceled")
			ui.CommandEnd(ui.StatusAbort)
			return nil
		}

		if choice == "set" {
			domain, err = ui.Input("Domain suffix (e.g., example.com)", "")
			if err != nil {
				ui.Abort("Canceled")
				ui.CommandEnd(ui.StatusAbort)
				return nil
			}
		}
	}

	// Credential
	username := flags.username
	var password string

	if flags.usernameSet {
		// Username provided via flag - validate and prompt for password
		if username == "" {
			ui.Error("Username cannot be empty")
			ui.CommandEnd(ui.StatusError)
			return nil
		}
		if !flags.dryRun {
			password, err = ui.PasswordWithConfirm("Password")
			if err != nil {
				ui.Error("Password error: %s", err)
				ui.CommandEnd(ui.StatusError)
				return nil
			}
		}
	} else if !flags.dryRun {
		// No username flag - offer interactive credential setup
		choice, err := ui.Select("Credential", []ui.SelectOption{
			{Label: "Skip (no credential)", Value: "skip"},
			{Label: "Add credential...", Value: "add"},
		})
		if err != nil {
			ui.Abort("Canceled")
			ui.CommandEnd(ui.StatusAbort)
			return nil
		}

		if choice == "add" {
			username, err = ui.Input("Username", "")
			if err != nil {
				ui.Abort("Canceled")
				ui.CommandEnd(ui.StatusAbort)
				return nil
			}
			if username == "" {
				ui.Warning("Username cannot be empty, skipping credential")
			} else {
				password, err = ui.PasswordWithConfirm("Password")
				if err != nil {
					ui.Error("Password error: %s", err)
					ui.CommandEnd(ui.StatusError)
					return nil
				}
			}
		}
	}

	if flags.dryRun {
		ui.SubSection("Preview")
		ui.Info("Would create context '%s' for file '%s'", finalName, gitIncludeFile)
		if domain != "" {
			ui.Info("Domain auto-select: %s", domain)
		}
		if username != "" {
			ui.Info("Would add credential for '%s'", username)
		}
		ui.Warning("Dry-run: no changes written")
		ui.CommandEnd(ui.StatusNoop)
		return nil
	}

	// Create SSH config file if needed
	paths := config.DefaultPaths()
	confD := filepath.Join(paths.SSHConfigDir, "conf.d")
	configFilePath := filepath.Join(confD, gitIncludeFile)
	fileCreated := false

	if _, err := os.Stat(configFilePath); os.IsNotExist(err) {
		if err := os.MkdirAll(confD, 0700); err != nil {
			ui.Error("Failed to create conf.d: %s", err)
			ui.CommandEnd(ui.StatusError)
			return nil
		}
		f, err := os.Create(configFilePath)
		if err != nil {
			ui.Error("Failed to create config file: %s", err)
			ui.CommandEnd(ui.StatusError)
			return nil
		}
		_ = f.Close()
		fileCreated = true
	}

	// Build credential if provided
	var cred *vault.Credential
	if username != "" && password != "" {
		cred = &vault.Credential{Username: username, Password: password}
	}

	// Create context with credential (single write)
	// Store just the filename - full path is constructed at runtime via sshConfigPath()
	if err := mgr.CreateContext(finalName, gitIncludeFile, domain, cred); err != nil {
		ui.Error("Failed to create context: %s", err)
		ui.CommandEnd(ui.StatusError)
		return nil
	}

	// Show results
	if fileCreated {
		ui.Success("Created SSH config file: '%s'", configFilePath)
	}
	ui.Success("Context '%s' created", finalName)
	if domain != "" {
		ui.Info("Auto-selects for hosts matching: *.%s", domain)
	}
	if username != "" {
		ui.Success("Credential added for '%s'", username)
	}

	ui.CommandEnd(ui.StatusSuccess)
	return nil
}

File: src/nssh/cli/ctx/add.py
"""Context add command for nssh ctx."""

from __future__ import annotations

from typing import Optional

from nssh.cli import click
from nssh.cli.common.banner import FAIL, NOOP, OK, banner
from nssh.cli.common.credentials import get_manager
from nssh.cli.common.prompt import (
    ask_text,
    prompt_password_with_confirmation,
    prompt_required,
)
from nssh.cli.common.ssh_include import ensure_conf_d_include
from nssh.core.env.paths import ssh_include_dir
from nssh.core.ui.console import get_console

console = get_console()


@click.command(short_help="Create a new context")
@click.argument("name", required=False, default=None)
@click.option("--domain", "-d", default=None, help="Domain suffix for auto-selection")
@click.option("--dry-run", is_flag=True, default=False, help="Preview only")
@click.pass_context
def add_command(
    ctx: click.Context, name: Optional[str], domain: Optional[str], dry_run: bool
) -> None:
    """Create new SSH context.

    Prompts for context name and SSH config file interactively.

    The --domain option enables automatic context selection when connecting
    to hosts matching the domain suffix (e.g., --domain example.com will
    auto-select this context for server.example.com).
    """
    cm = get_manager(ctx)

    # Ensure the include wiring exists so the context file will be loaded
    ensure_conf_d_include(
        create_if_missing=True,
        abort_on_decline=True,
        preview_title="SSH config change preview",
        dry_run=dry_run,
    )

    with banner("CREATE NEW SSH CONTEXT", OK) as set_outcome:
        _add_context(cm, name, domain, dry_run, set_outcome)


def _add_context(cm, name, domain, dry_run, set_outcome) -> None:
    """Internal implementation for adding a context."""
    # Interactive prompts
    if dry_run:
        final_name = name or "<dry-run-context>"
        git_include_file = "<dry-run-file>"
    else:
        final_name = prompt_required("Context name", name)

        # Check if context already exists
        existing = [c.get("name") for c in cm.list_contexts()]
        if final_name in existing:
            console.print(f"[red]Error: Context '{final_name}' already exists[/red]")
            set_outcome(FAIL)
            raise SystemExit(1)

        conf_d = ssh_include_dir()
        default_filename = f"{final_name}_hosts"
        git_include_file = ask_text(
            f"SSH config filename (in {conf_d}/)",
            default=default_filename,
            allow_empty=False,
        )

    # Prompt for optional credentials (before any mutations)
    username = ask_text(
        "Username (optional)",
        default="",
        allow_empty=True,
    )
    password = None
    if username:
        try:
            password = prompt_password_with_confirmation(
                "Password",
                allow_empty=False,
            )
        except ValueError as exc:
            console.print(f"[red]Error: {exc}[/red]")
            set_outcome(FAIL)
            raise SystemExit(1)

    if dry_run:
        console.print()
        console.print(
            f"[dim]Would create context '{final_name}' for file '{git_include_file}'[/dim]"
        )
        if domain:
            console.print(f"[dim]Domain auto-select: {domain}[/dim]")
        if username:
            console.print(f"[dim]Would add credential for '{username}'[/dim]")
        console.print("[dim]Dry-run: no changes written[/dim]")
        set_outcome(NOOP)
        return

    try:
        # Check if the SSH config file exists, create if needed
        conf_d = ssh_include_dir()
        config_file_path = conf_d / git_include_file
        file_created = False
        if not config_file_path.exists():
            conf_d.mkdir(parents=True, exist_ok=True)
            config_file_path.touch()
            file_created = True

        cm.create_context(final_name, git_include_file, domain=domain)

        # Add credential if provided
        if username and password:
            cm.add_context_credential(final_name, username, password, overwrite=True)

        # Show all results together
        console.print()
        if file_created:
            console.print(
                f"[green]+[/green] Created SSH config file: '{config_file_path}'"
            )
        console.print(f"[green]+[/green] Context '{final_name}' created")
        if domain:
            console.print(f"  Auto-selects for hosts matching: *.{domain}")
        if username:
            console.print(f"[green]+[/green] Credential added for '{username}'")
        # Default footer "Context created" will be used
    except Exception as exc:
        console.print(f"[red]Error: {exc}[/red]")
        set_outcome(FAIL)
        raise SystemExit(1)
